### PR TITLE
Feat/2.5 supply and borrow caps: aggregated

### PR DESCRIPTION
**Pull request by dhadrien**

### CODE REVIEW

**Comment by dhadrien** [comment]
#82 
#94

File: contracts/protocol/lendingpool/LendingPoolConfigurator.sol [hunk 1]
```diff
@@ -248,13 +248,14 @@ contract LendingPoolConfigurator is VersionedInitializable, ILendingPoolConfigur
...
    **/
-  function enableBorrowingOnReserve(address asset, bool stableBorrowRateEnabled)
+  function enableBorrowingOnReserve(address asset, uint256 borrowCap, bool stableBorrowRateEnabled)
     external
     onlyPoolAdmin
   {
     DataTypes.ReserveConfigurationMap memory currentConfig = pool.getConfiguration(asset);
 
     currentConfig.setBorrowingEnabled(true);
+    currentConfig.setBorrowCap(borrowCap);
```
Comment (The-3D) [inline contracts/protocol/lendingpool/LendingPoolConfigurator.sol:258]: If you are setting the borrow cap, you also need to emit a ```BorrowCapChanged```

File: contracts/protocol/libraries/configuration/ReserveConfiguration.sol [hunk 2]
```diff
@@ -19,6 +19,8 @@ library ReserveConfiguration {
   uint256 constant BORROWING_MASK =             0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFBFFFFFFFFFFFFFF; // prettier-ignore
   uint256 constant STABLE_BORROWING_MASK =      0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF7FFFFFFFFFFFFFF; // prettier-ignore
   uint256 constant RESERVE_FACTOR_MASK =        0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF0000FFFFFFFFFFFFFFFF; // prettier-ignore
+  uint256 constant BORROW_CAP_MASK =            0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF000000000000FFFFFFFFFFFFFFFFFFFF; // prettier-ignore
```
Comment (The-3D) [inline contracts/protocol/libraries/configuration/ReserveConfiguration.sol:22]: reduce the size of the fields for borrow cap and supply cap to 36 bit. 36 bit can still hold 68 billions, which should be more than enough to handle a supply/borrow cap. This will allow to save 24 bits for future fields

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 3]
```diff
@@ -44,6 +46,13 @@ library ValidationLogic {
     require(amount != 0, Errors.VL_INVALID_AMOUNT);
     require(isActive, Errors.VL_NO_ACTIVE_RESERVE);
     require(!isFrozen, Errors.VL_RESERVE_FROZEN);
+    require(
+      IERC20(reserve.aTokenAddress)
+        .totalSupply()
+        .add(amount)
+        .div(10 ** reserve.configuration.getDecimals()) < reserve.configuration.getSupplyCap(),
```
Comment (The-3D) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:53]: cache reserve.configuration in a local variable to avoid multiple SLOAD

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 4]
```diff
@@ -126,6 +137,22 @@ library ValidationLogic {
         uint256(DataTypes.InterestRateMode.STABLE) == interestRateMode,
       Errors.VL_INVALID_INTEREST_RATE_MODE_SELECTED
     );
+    
+    (vars.totalSupplyStableDebt, ) = IStableDebtToken(reserve.stableDebtTokenAddress)
```
Comment (The-3D) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:141]: use totalSupply()

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 5]
```diff
@@ -126,6 +137,22 @@ library ValidationLogic {
...
+      .getTotalSupplyAndAvgRate();
+
+    vars.totalSupplyVariableDebt = IVariableDebtToken(reserve.variableDebtTokenAddress)
```
Comment (The-3D) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:144]: use totalSupply(). We will refactor the code afterwards to fix #122 and replace this with a cached version

File: contracts/protocol/libraries/types/DataTypes.sol [hunk 6]
```diff
@@ -38,6 +38,8 @@ library DataTypes {
     //bit 59: stable rate borrowing enabled
     //bit 60-63: reserved
     //bit 64-79: reserve factor
+    //bit 80-127 borrow cap
```
Comment (The-3D) [inline contracts/protocol/libraries/types/DataTypes.sol:41]: change this comment once you change the field size

File: helpers/constants.ts [hunk 7]
```diff
@@ -15,6 +15,8 @@ export const oneEther = new BigNumber(Math.pow(10, 18));
 export const oneRay = new BigNumber(Math.pow(10, 27));
 export const MAX_UINT_AMOUNT =
   '115792089237316195423570985008687907853269984665640564039457584007913129639935';
+export const MAX_BORROW_CAP = '281474976710655';
```
Comment (The-3D) [inline helpers/constants.ts:18]: change this once you change the field size

File: contracts/protocol/libraries/configuration/ReserveConfiguration.sol [hunk 8]
```diff
@@ -29,12 +31,16 @@ library ReserveConfiguration {
...
   uint256 constant RESERVE_FACTOR_START_BIT_POSITION = 64;
+  uint256 constant BORROW_CAP_START_BIT_POSITION = 80;
+  uint256 constant SUPPLY_CAP_START_BIT_POSITION = 128;
 
   uint256 constant MAX_VALID_LTV = 65535;
   uint256 constant MAX_VALID_LIQUIDATION_THRESHOLD = 65535;
   uint256 constant MAX_VALID_LIQUIDATION_BONUS = 65535;
   uint256 constant MAX_VALID_DECIMALS = 255;
   uint256 constant MAX_VALID_RESERVE_FACTOR = 65535;
+  uint256 constant MAX_VALID_BORROW_CAP = 281474976710655;
```
Comment (The-3D) [inline contracts/protocol/libraries/configuration/ReserveConfiguration.sol:42]: change this once you change the field size

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 9]
```diff
@@ -44,6 +46,13 @@ library ValidationLogic {
     require(amount != 0, Errors.VL_INVALID_AMOUNT);
     require(isActive, Errors.VL_NO_ACTIVE_RESERVE);
     require(!isFrozen, Errors.VL_RESERVE_FROZEN);
+    require(
+      IERC20(reserve.aTokenAddress)
```
Comment (The-3D) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:50]: change this condition: check if the supply cap is 0 (unset) || total supply < supply cap

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 10]
```diff
@@ -126,6 +137,22 @@ library ValidationLogic {
...
+    (vars.totalSupplyStableDebt, ) = IStableDebtToken(reserve.stableDebtTokenAddress)
+      .getTotalSupplyAndAvgRate();
+
+    vars.totalSupplyVariableDebt = IVariableDebtToken(reserve.variableDebtTokenAddress)
+          .scaledTotalSupply()
+          .rayMul(reserve.variableBorrowIndex);
+        
+      
+    require(
+      vars.totalSupplyStableDebt
```
Comment (The-3D) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:150]: change this condition: check getBorrowCap() == 0 || total debt < borrow cap

File: markets/aave/reservesConfigs.ts [hunk 11]
```diff
@@ -21,7 +22,9 @@ export const strategyBUSD: IReserveParams = {
   stableBorrowRateEnabled: false,
   reserveDecimals: '18',
   aTokenImpl: eContractid.AToken,
-  reserveFactor: '1000'
+  reserveFactor: '1000',
+  borrowCap: MAX_BORROW_CAP,
```
Comment (The-3D) [inline markets/aave/reservesConfigs.ts:26]: once you change the conditions in validationLogic, you can leave this unset

File: markets/amm/reservesConfigs.ts [hunk 12]
```diff
@@ -15,7 +16,9 @@ export const strategyWETH: IReserveParams = {
   stableBorrowRateEnabled: false,
   reserveDecimals: '18',
   aTokenImpl: eContractid.AToken,
-  reserveFactor: '1000'
+  reserveFactor: '1000',
+  borrowCap: MAX_BORROW_CAP,
```
Comment (The-3D) [inline markets/amm/reservesConfigs.ts:20]: same as the main market

File: markets/matic/reservesConfigs.ts [hunk 13]
```diff
@@ -20,6 +21,8 @@ export const strategyDAI: IReserveParams = {
   reserveDecimals: '18',
   aTokenImpl: eContractid.AToken,
   reserveFactor: '1000',
+  borrowCap: MAX_BORROW_CAP,
```
Comment (The-3D) [inline markets/matic/reservesConfigs.ts:24]: same as the main market

File: package.json [hunk 14]
```diff
@@ -149,7 +149,8 @@
     "Andrey Kozlov <andrey@aave.com>",
     "David Racero <david.k@aave.com>",
     "Pol Sendra <pol@aave.com>",
-    "David Truong <david@aave.com>"
+    "David Truong <david@aave.com>",
+    "Hadrien Charlanes <hadrien@aave.com>"
```
Comment (The-3D) [inline package.json:153]: create a separate PR next time

File: test-suites/test-aave/configurator.spec.ts [hunk 15]
```diff
@@ -355,15 +408,82 @@ makeSuite('LendingPoolConfigurator', (testEnv: TestEnv) => {
     expect(liquidationThreshold).to.be.equal(strategyWETH.liquidationThreshold);
     expect(liquidationBonus).to.be.equal(strategyWETH.liquidationBonus);
     expect(stableBorrowRateEnabled).to.be.equal(strategyWETH.stableBorrowRateEnabled);
+    expect(borrowCap).to.be.equal(strategyWETH.borrowCap);
+    expect(supplyCap).to.be.equal(strategyWETH.supplyCap);
     expect(reserveFactor).to.be.equal(1000);
   });
 
-  it('Check the onlyLendingPoolManager on setReserveFactor', async () => {
+  it('Fails to change to too high borrowCap', async () => {
```
Comment (The-3D) [inline test-suites/test-aave/configurator.spec.ts:416]: Change title of the test - "Check that a borrow cap that exceeds the MAX_BORROW_CAP cannot be set"

File: test-suites/test-aave/configurator.spec.ts [hunk 16]
```diff
@@ -355,15 +408,82 @@ makeSuite('LendingPoolConfigurator', (testEnv: TestEnv) => {
...
     const { configurator, users, weth } = testEnv;
     await expect(
-      configurator.connect(users[2].signer).setReserveFactor(weth.address, '2000'),
+      configurator.setBorrowCap(weth.address, BigNumber.from(MAX_BORROW_CAP).add(1)),
       CALLER_NOT_POOL_ADMIN
-    ).to.be.revertedWith(CALLER_NOT_POOL_ADMIN);
+    ).to.be.revertedWith(RC_INVALID_BORROW_CAP);
+  });
+  it('Fails to change to too high supplyCap', async () => {
```
Comment (The-3D) [inline test-suites/test-aave/configurator.spec.ts:423]: same as the other

File: test-suites/test-aave/configurator.spec.ts [hunk 17]
```diff
@@ -355,15 +408,82 @@ makeSuite('LendingPoolConfigurator', (testEnv: TestEnv) => {
...
+    expect(ltv).to.be.equal(strategyWETH.baseLTVAsCollateral);
+    expect(liquidationThreshold).to.be.equal(strategyWETH.liquidationThreshold);
+    expect(liquidationBonus).to.be.equal(strategyWETH.liquidationBonus);
+    expect(stableBorrowRateEnabled).to.be.equal(strategyWETH.stableBorrowRateEnabled);
+    expect(reserveFactor).to.be.equal(1000);
+    expect(borrowCap).to.be.equal('3000000');
+    expect(supplyCap).to.be.equal(strategyWETH.supplyCap);
+  });
+
+  it('Changes the borrow Cap of WETH', async () => {
```
Comment (The-3D) [inline test-suites/test-aave/configurator.spec.ts:460]: this one changes the supply cap

File: test-suites/test-amm/configurator.spec.ts [hunk 18]
```diff
@@ -178,7 +178,7 @@ makeSuite('LendingPoolConfigurator', (testEnv: TestEnv) => {
     expect(ltv).to.be.equal(strategyWETH.baseLTVAsCollateral);
     expect(liquidationThreshold).to.be.equal(strategyWETH.liquidationThreshold);
     expect(liquidationBonus).to.be.equal(strategyWETH.liquidationBonus);
-    expect(stableBorrowRateEnabled).to.be.equal(true/*strategyWETH.stableBorrowRateEnabled*/);
```
Comment (The-3D) [inline test-suites/test-amm/configurator.spec.ts:181]: why this change?

**Review by The-3D — APPROVED** [review]
